### ISSUE TITLE
Problems with ANSI sequences in cmd.exe

### ISSUE THREAD

**Issue by viviwon** [sharkdp/numbat#291]
like this 
![image](https://github.com/sharkdp/numbat/assets/153968785/b5ffd7b9-4abb-4af9-921a-c6e4cffd5b4f)
I do love this tool, so can this software be packaged as a Docker image and deployed on my own? Sometimes web access is not very smooth

**Comment by eminence** [comment]
Your terminal isn't handling ANSI/xterm color escape codes.   What terminal are you using and what version of Windows are you running?

In Windows 10/11, both [Windows Terminal](https://apps.microsoft.com/detail/9N0DX20HK701?hl=en-US&gl=US) and cmd.exe should support this

**Comment by viviwon** [comment]
> Your terminal isn't handling ANSI/xterm color escape codes. What terminal are you using and what version of Windows are you running?
> 
> In Windows 10/11, both [Windows Terminal](https://apps.microsoft.com/detail/9N0DX20HK701?hl=en-US&gl=US) and cmd.exe should support this

My computer version is Windows 10 Home Chinese Edition 1909. I'm not sure which terminal I am using, but I run the software directly from its .exe file. It's strange that I could use it normally yesterday, but today when I opened my computer, it's not working. I don't know the reason behind it.

**Comment by eminence** [comment]
According to [this page](https://ss64.com/nt/syntax-ansi.html) your version (1909) should support this.

> It's strange that I could use it normally yesterday, but today when I opened my computer, it's not working

What changed between yesterday and today?  Are you using the same version of numbat?

**Comment by viviwon** [comment]
> According to [this page](https://ss64.com/nt/syntax-ansi.html) your version (1909) should support this.
> 
> > It's strange that I could use it normally yesterday, but today when I opened my computer, it's not working
> 
> What changed between yesterday and today? Are you using the same version of numbat?

I was previously using version 1.8.0, and I always had this issue, so I stopped using it. Yesterday, I switched to version 1.9.0, and it worked, but today it's not working again. I'm not sure why because I haven't made any changes.

**Comment by sharkdp** [comment]
Maybe it has to do with encodings (since you mentioned *Chinese* above)? I remember seeing similar problems in another program of mine when variable-width encodings were used in the terminal.

As a workaround, you could try to set the `NO_COLOR` environment variable to an arbitrary value. This will disable colorization via ANSI escape sequences.

**Comment by viviwon** [comment]
> Maybe it has to do with encodings (since you mentioned _Chinese_ above)? I remember seeing similar problems in another program of mine when variable-width encodings were used in the terminal.
> 
> As a workaround, you could try to set the `NO_COLOR` environment variable to an arbitrary value. This will disable colorization via ANSI escape sequences.

Thank you very much! I tried it, and it indeed works. However, I need to enter "let no_color=1" every time

**Comment by sharkdp** [comment]
> However, I need to enter "let no_color=1" every time

wait, what? :-)

you're supposed to set the variable in your shell. not in numbat

**Comment by viviwon** [comment]
> you're supposed to set the variable in your shell. not in numbat

I'm sorry, I'm not very knowledgeable about computers. I don't know how to make the change you suggested. However, I found that if I open the .exe file and enter a random word before continuing to use it, the issue is resolved.

**Comment by eminence** [comment]
> However, I found that if I open the .exe file and enter a random word before continuing to use it, the issue is resolved.

Than is strange, I cannot explain that.

In a cmd.exe window, can you try to run this before running numbat.exe:

```
set NO_COLOR-1
```

It will look like this:

![image](https://github.com/sharkdp/numbat/assets/402454/4a393720-dfd4-4129-bf5a-5810cc9b04fe)

**Comment by viviwon** [comment]
> In a cmd.exe window, can you try to run this before running numbat.exe:
![image](https://github.com/sharkdp/numbat/assets/153968785/e1a2715f-1f96-4613-aa9e-38d4f557ef1b)

That does work properly. This is how I directly run numbat.exe.

![image](https://github.com/sharkdp/numbat/assets/153968785/2a2fdecb-799f-4f3f-982f-3623e206be76)

**Comment by eminence** [comment]
Sorry, I cannot explain why the colors in the first expression aren't rendering properly.  This will be very hard for us to debug, unless one of us can reproduce it locally (which I can't).

@viviwon now that you have a workaround, is there anything else we can help with?

**Comment by sharkdp** [comment]
I'm closing this due to inactivity. Please feel free to comment in case it should be re-opened.

**Comment by Turun** [comment]
Ok, I did some digging.

The issue is that Powershell (and cmd. I write ps, but it applies to both as far as I can tell) is by default not in [ENABLE_VIRTUAL_TERMINAL_PROCESSING mode](https://learn.microsoft.com/en-us/windows/console/setconsolemode). Big discussion if that should be changed here: https://github.com/PowerShell/PowerShell/issues/19101  
But powershell commands themselves temporarily enable VT Codes so they can look pretty. So it may not be immediately obvious that the terminal does not handle color codes. This can be seen by comparing the output of e.g. `python -c 'print(''It ain\''t easy being \x1b[32mgreen\x1b[m.'')'` with `(python -c 'print(''It ain\''t easy being \x1b[32mgreen\x1b[m.'')')`. The latter one has its output buffered by powershell (this is what the parenthesis around the command do) and so will appear in color, whereas the first one will print the escape codes like the OP of this issue. Users can also manually change a setting/registry key to change the default from disabled to enabled. I believe the Windows Terminal does this, though I can't check that here.

Programs can also enable and disable this mode as desired. This is something that cargo does when compiling, so if you clone the repo and do `cargo run` the issue can not be reproduced! Because numbat is called after cargo told the terminal to process VT Codes and before powershell reset the mode back to the default disabled. To reproduce locally you need to have the default settings (no registry key or changed default) and do `cargo build` and then `./target/debug/numbat.exe` (or `cargo install -f --path numbat-cli` and then `numbat`)!

# Why is the problem being fixed when triggering an error?  

Because errors are printed with
```rust
pub fn print_diagnostic(&self, error: impl ErrorDiagnostic) {
    use codespan_reporting::term::{
        self,
        termcolor::{ColorChoice, StandardStream},
        Config,
    };

    let writer = StandardStream::stderr(ColorChoice::Auto);
    let config = Config::default();

    // we want to be sure no one can write between our diagnostics
    let mut writer = writer.lock();
    for diagnostic in error.diagnostics() {
        term::emit(&mut writer, &config, &self.resolver.files, &diagnostic).unwrap();
    }
}
```
whereas the usual output is printed with e.g.
```rust

let help = help_markup();
print!("{}", ansi_format(&help, true));
```

`termcolor` sets the ENABLE_VIRTUAL_TERMINAL_PROCESSING flag somewhere in its code. Unfortunately I have not figured out where/when/how. If the `ansi_format` output is printed before that, the terminal does not yet have this flag enabled and therefore prints the characters directly.

# The solution is to either

a) print everything with `termcolor`, which would ensure the flag is enabled on every print call

b) print something early on once with `termcolor`, for example the logo/version number which is the first thing the program outputs to the screen. Thereby ensuring the flag is set for the subsequent text which is printed with `ansi_format`

c) explicitly enable the flag in the code. For example by using    
```rust
use console;  // new dependency, so maybe not preferred
console::set_colors_enabled(true);
```
before entering the repl loop.

**Comment by sharkdp** [comment]
> `termcolor` sets the ENABLE_VIRTUAL_TERMINAL_PROCESSING flag somewhere in its code. Unfortunately I have not figured out where/when/how.

I'm not sure, but I think termcolor does not use ANSI colors for colorization on Windows. Anyway, the rest of your analysis is correct — thank you. Should be fixed by #670

**Comment by sharkdp** [comment]
This should be fixed in v1.15.

**Comment by Turun** [comment]
I can confirm that the Problem is fixed for me 👍 (checked on current master, aa96754)
